### ISSUE TITLE
Possible soundness issue related to use-after-free or lifetimes when storing a reference in the FileSystemContext

### ISSUE THREAD

**Issue by drey7925** [SnowflakePowered/winfsp-rs#4]
With [this gist](https://gist.github.com/drey7925/7100fdda890fd41ef102f15254c1aad2) containing only Safe Rust code, it appears to be possible to provoke a use-after-free with winfsp. 

The intent of the gist is to provide a minimal WinFSP filesystem that I can see in "My Computer" that includes enough structure to reproduce my finding; each time I check the volume properties, the free space should increase slightly and the volume label should change.

When the code in the gist is run, the first three lines are:

```
Started FSP instance
dropping!
The service bugreprofs has been started.
```

showing that the FakeFsBackend gets dropped while the filesystem still has a reference to it. At the same time, I observed a probability of causing an access violation, or overwriting the atomic integer (which is visible by the free space and volume label being unusually large), thus confirming likely memory corruption.

Could you take a look and let me know if this is a bug, or if perhaps I'm violating some extra invariant that winfsp-rs expects? I'd expect that safe code shouldn't be able to cause a soundness issue in this way, but I may be mistaken as I'm a relative beginner with Rust.

Thanks in advance!

**Comment by chyyran** [comment]
Hi, sorry for the delayed response. My GitHub notifications have not been working lately. 

I think you're correct that this is a soundness bug. I'll have to look more into it, thanks for the issue report.

**Comment by chyyran** [comment]
Fixed in #6

This was a breaking API change to add a lifetime parameter to `FileSystemHost`. The example you listed should no longer compile with 

```
error[E0515]: cannot return value referencing local variable `backend`                                                                                                                                                                                                                                                                       
  --> unsound-fs\src\main.rs:99:13
   |
95 |             let mut fs = FakeFs::create(&backend).unwrap();
   |                                         -------- `backend` is borrowed here
...
99 |             Ok(fs)
   |             ^^^^^^ returns a value referencing data owned by the current function
```

The new API in `0.6.x` specifies an explicit lifetime parameter `'ctx` for `FileSystemHost`. I've tried to fix this in an as-unbreaking manner as possible in `0.5.x` by making this lifetime `'static`.
